### ISSUE TITLE
tensor infer_shape

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#476]
There is many op under tensor that don't implement infer_shape. Also, those that do, do not always test it or don't test it correctly.

TODO: implement missing infer_shape and add tests of them with the utt.InferShapeTester class.
TODO: add missing/move infer_shape test to use the utt.InferShapeTester system.

file tensor/basic.py Op without infer_shape
TensorDot, TensorDotGrad, AdvancedIncSubtensor, PermuteRowElements, Tile, Flatten, Split, Default, Eye, Shape, ScalarFromTensor, TensorFromScalar
with infer_shape:Dot, AdvancedSubtensor, AdvancedIncSubtensor1, AdvancedSubtensor1, ARange, Reshape,Join,Rebroadcast, IncSubtensor, Subtensor, Alloc, MaxAndArgmax, SpecifyShape, Mean(CAReduce)

tensor/blas{,_c,_scipy}.py: op that don't need infer_shape(the are added by optimization) (NOTHING TODO)

tensor/opt.py: without infer_shape MakeVector(also move it to basic.py to don't have basic.py depend on opt.py), Shape_i,
with infer_shape: Assert

tensor/elemwise.py: 
with infer_shape Dimshuffle, Elemwise, CAReduce

tensor/nnet/ConvGrad3D.py, with infer_shape ConvGrad3D
tensor/nnet/Conv3D.py, with infer_shape Conv3D
tensor/nnet/ConvTransp3D.py, with infer_shape ConvTransp3D
tensor/nnet/conv.py  with infer_shape ConOp
tensor/nnet/nnet.py  with infer_shape SoftmaxWithBias, SoftmaxGrad, Softmax, CrossentropySoftmaxArgmax1HotWithBias, CrossentropySoftmax1HotWithBiasDx
without infer_shape: CrossentropyCategorical1HotGrad, CrossentropyCategorical1Hot, Prepend_scalar_constant_to_each_row, Prepend_scalar_to_each_row

tensor/signal/downsample.py witout infer_shape: DownsampleFactorMax, DownsampleFactorMaxGrad

tensor/raw_random.py: with infer_shape: RandomFunction

**Comment by nouiz** [comment]
I think there is only one left and it have its own ticket gh-788. So I close this one.
